### PR TITLE
feat(EMS-805): Your buyer - Company or organisation - continue and save and go back

### PR DESCRIPTION
**Pull request by Zainzzkk**
# Changes made:
* Added buyer table to database
* Added save functionality for buyer
* Added save and go back functionality
* Added tests
* Added e2e tests
* Restructured your buyer section in ui folder and e2e test folder
* Fixed bug when country would be deleted from select if validation errors (EMS-1105)
* Updated UI rate limiter to only run in production - this was causing issues in GHA E2E tests.

### CODE REVIEW

File: src/api/schema.ts [hunk 1]
```diff
@@ -216,6 +228,18 @@ export const lists = {
...
+            await context.db.Buyer.updateOne({
+              where: { id: buyerId },
+              data: {
+                application: {
+                  connect: {
+                    id: applicationId,
+                  },
+                },
+              },
+            });
```
Comment (ttbarnes) [inline src/api/schema.ts:242]: Can we add unit test coverage for the new relationships please, in `keystone.test.ts`

File: e2e-tests/cypress/e2e/journeys/insurance/your-buyer/company-or-organisation/save-and-back.spec.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      cy.checkValue(companyOrOrganisationPage[REGISTRATION_NUMBER], '');
+      cy.checkValue(companyOrOrganisationPage[WEBSITE], '');
+      cy.checkValue(companyOrOrganisationPage[FIRST_NAME], '');
+      cy.checkValue(companyOrOrganisationPage[LAST_NAME], '');
+      cy.checkValue(companyOrOrganisationPage[POSITION], '');
+      cy.checkValue(companyOrOrganisationPage[EMAIL], '');
+    });
+  });
+
+  describe('save and back on a fully entered form', () => {
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-buyer/company-or-organisation/save-and-back.spec.js:112]: In the other test this is described as 

```
describe('when submitting a fully filled form', () => {
```

Can we keep them the same?

File: src/api/schema.ts [hunk 3]
```diff
@@ -404,6 +428,30 @@ export const lists = {
...
+      contactFirstName: text(),
+      contactLastName: text(),
+      contactPosition: text(),
+      contactEmail: text(),
+      canContactBuyer: select({
+        options: [
+          { label: ANSWERS.YES, value: ANSWERS.YES },
+          { label: ANSWERS.NO, value: ANSWERS.NO },
+        ],
+        db: { isNullable: true },
```
Comment (ttbarnes) [inline src/api/schema.ts:450]: If this is a required field. I think we should/could remove this line

File: src/ui/server/api/keystone/application/index.ts [hunk 4]
```diff
@@ -215,6 +216,36 @@ const application = {
...
+
+        const variables = {
+          where: { id },
+          data: update,
+        };
+
+        const response = (await apollo('POST', updateBuyerMutation, variables)) as ApolloResponse;
+
+        if (response.errors) {
+          console.error('GraphQL error updating application buyer', response.errors);
```
Comment (ttbarnes) [inline src/ui/server/api/keystone/application/index.ts:231]: Can we add a space at the end of the message, and in the `console.error` below, just to be consistent

File: src/ui/server/controllers/insurance/your-buyer/map-and-save/index.API-ERROR.test.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import mapAndSave from '.';
+import save from '../save-data';
+import { mockApplication } from '../../../../test-mocks';
+
+describe('controllers/insurance/business/map-and-save - api errors', () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/map-and-save/index.API-ERROR.test.ts:5]: Can we lowercase the `API-ERROR` part of the file name? Noticed another file has the same name

File: src/ui/server/controllers/insurance/your-buyer/map-and-save/index.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * maps buyer request and calls save function
+ * returns true or false based on response from save function
+ * @param {RequestBody} formBody
+ * @param {Object} application
+ * @param {Object} validationErrors
+ * @returns {Boolean}
+ */
+const buyer = async (formBody: RequestBody, application: Application, validationErrors?: ValidationErrors) => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/map-and-save/index.ts:13]: should we call it `yourBuyer` ?

File: src/ui/server/controllers/insurance/your-buyer/save-data/index.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { sanitiseData } from '../../../../helpers/sanitise-data';
+import { Application, RequestBody } from '../../../../../types';
+
+/**
+ * gets fields to add to the database and sanitises them
+ * saves to buyer table in database via api call
+ * @param {Application} application
+ * @param {RequestBody} formBody
+ * @param {Object} errorList
+ * @returns {Object} saveResponse from api
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/save-data/index.ts:13]: Can we uppercase `api` to be consistent

File: src/ui/server/helpers/sanitise-data/index.test.ts [hunk 8]
```diff
@@ -53,6 +56,12 @@ describe('server/helpers/sanitise-data', () => {
         ADDRESS_LINE_2,
         POSTCODE,
         BROKER_NAME,
+        NAME,
+        REGISTRATION_NUMBER,
+        ADDRESS,
+        FIRST_NAME,
+        LAST_NAME,
+        POSITION,
```
Comment (ttbarnes) [inline src/ui/server/helpers/sanitise-data/index.test.ts:64]: Should `POSITION` allow pure numbers?

File: src/ui/server/test-mocks/mock-application.ts [hunk 9]
```diff
@@ -5,6 +5,7 @@ import mockCountries from './mock-countries';
 import mockCurrencies from './mock-currencies';
 import { Application } from '../../types';
 import mockBroker from './mock-broker';
+import mockBuyerExample from './mock-buyer';
```
Comment (ttbarnes) [inline src/ui/server/test-mocks/mock-application.ts:8]: `mockBuyer`

Comment (Zainzzkk) [reply src/ui/server/test-mocks/mock-application.ts:8]: used mockBuyerExample as mockBuyer is being used below

File: src/ui/server/helpers/sanitise-data/index.test.ts (hunk 8 above)
Comment (Zainzzkk) [reply src/ui/server/helpers/sanitise-data/index.test.ts:64]: technically could be an edge case which would break the db call

File: src/ui/server/test-mocks/mock-application.ts (hunk 9 above)
Comment (ttbarnes) [reply src/ui/server/test-mocks/mock-application.ts:8]: Maybe we can do this - rename `mockBuyerExample` to `mockBuyer`, and rename line 81 to `mockApplicationBuyer`?

**Review by ttbarnes — APPROVED** [review]
